src/foam/nanos/auth/AuthenticatedGroupDAO.java

```diff
@@ -0,0 +1,76 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.FObject;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.dao.ProxyDAO;
+import foam.util.SafetyUtil;
+
+/**
+ * Custom authentication for groupDAO
+ *
+ * Features:
+ *   - If creating a group, you can only give it permissions implied by those
+ *     that you already have.
+ *   - If updating a group, you can't modify the permissions array to have a
+ *     permission that isn't implied by one you have.
+ *   - If updating a group, you can't modify the parent of the group such that
+ *     the group gains a permission that isn't implied by one you have.
+ */
+public class AuthenticatedGroupDAO extends ProxyDAO {
+  public AuthenticatedGroupDAO(X x, DAO delegate) {
+    super(x, delegate);
+  }
+
+  /**
+   * Make sure that the user has all permissions in the given group.
+   */
+  public void checkUserHasAllPermissionsInGroup(X x, Group toCheck) {
+    AuthService auth = (AuthService) x.get("auth");
+    String groupId;
+    Group group = toCheck;
+
+    while ( true ) {
```

Comment:
You should have a better condition. EG: `while( getAncestor(group) != null )` where `getAncestor(Group group)` is a helper that returns a groups if it exists null otherwise.

Reply:
Agree. You only have one break condition in your while loop (excluding exceptions) so you could make that your while loop exit condition.